torch/_inductor/pattern_matcher.py

```diff
@@ -2331,3 +2350,153 @@ def extract_target(node: torch.fx.Node) -> torch.fx.node.Target:
+    """
+    A no-op function that serves to create dependencies in the FX graph.
+    """
+    return t.clone()
+
+
+DEP_OP = torch.ops.pattern_matcher.op_for_dependencies
+
+
+def check_mutable_custom_op(graph: torch.fx.Graph) -> bool:
+    """Check if graph contains mutable operations or auto_functionalized nodes."""
+    for node in graph.nodes:
+        if node.op == "call_function" and (
+            (isinstance(node.target, torch._ops.OpOverload) and is_mutation_op(node))
+            or node.target
+            in (
+                torch.ops.higher_order.auto_functionalized,
+                torch.ops.higher_order.auto_functionalized_v2,
+            )
+        ):
+            return True
+    return False
+
+
+def add_implict_edges(gm: torch.fx.GraphModule) -> None:
+    """
+    Add implicit edges to preserve ordering of mutations.
+    """
+    from torch._higher_order_ops.auto_functionalize import get_mutable_args_from_schema
+    from torch.fx.passes.fake_tensor_prop import FakeTensorProp
+
+    from .fx_passes.memory_estimator import GraphAliasTracker, StorageKey
+
+    # Ensure FakeTensor metadata exists for placeholders so alias tracking works
+    if not any(
+        node.meta.get("val") is not None
+        for node in gm.graph.nodes
+        if node.op == "placeholder"
+    ):
+        FakeTensorProp(gm).propagate()
+
+    nodes = list(gm.graph.nodes)
+    alias_info = GraphAliasTracker(nodes)
+
+    # Build mapping from storage -> nodes that produce that storage as outputs
+    storage_to_nodes: dict[StorageKey, OrderedSet[torch.fx.Node]] = {}
+    for n in nodes:
+        for sk in alias_info.node_to_output_storages.get(n, OrderedSet()):
+            storage_to_nodes.setdefault(sk, OrderedSet()).add(n)
+
+    def mutated_arg_nodes(node: torch.fx.Node) -> list[torch.fx.Node]:
+        if not is_mutation_op(node) or not hasattr(node.target, "_schema"):
+            return []
+
+        mutable_names, _ = get_mutable_args_from_schema(node.target._schema)
+        if not mutable_names:
+            return []
+
+        mutated: list[torch.fx.Node] = []
+        for idx, arg in enumerate(node.target._schema.arguments):
+            if arg.name in mutable_names:
+                val = (
+                    node.args[idx]
+                    if idx < len(node.args)
+                    else node.kwargs.get(arg.name)
+                )
+                if isinstance(val, torch.fx.Node):
+                    mutated.append(val)
+        return mutated
+
+    def topo_index() -> dict[torch.fx.Node, int]:
+        return {n: i for i, n in enumerate(gm.graph.nodes)}
+
+    for node in list(gm.graph.nodes):
+        if node.op == "call_function" and node.target == DEP_OP:
+            continue
+        mutated_nodes = mutated_arg_nodes(node)
+        if not mutated_nodes:
+            continue
+
+        # For each mutated arg node, collect all storages it outputs
+        storage_to_mutated: dict[StorageKey, torch.fx.Node] = {}
```

Comment:
do we have the invariant that all mutable ops will return the tensor it mutates? should we assert this ? Otherwise we are just skipping which could be incorrect